### ISSUE TITLE
Compound tokens based on regular expressions including spaces within parentheses

### ISSUE THREAD

**Issue by jogrue** [quanteda/quanteda#1670]
## Describe the bug

As I mentioned in #1668, I have a dictionary where I switched from glob to regular expressions. I try to build still somewhat readable regular expressions using many parentheses for different evaluations. The dictionary also includes a list of multi-word compounds. On example would be this pattern (translated from the original German one): `"the (lower )99( |-)?(percent|%)"`

When trying to produce the compound tokens, such patterns fail. The problem is that the pattern is split at the blanks, leading to `"the" "(lower" ")?99(" "|-)?percent|%)"`. In search_regex_multi this leads to an error, because it then tries to match each sub pattern with `stri_detect_regex` and this lead to an `U_REGEX_MISMATCHED_PAREN` error. I do not think that there is an easy solution for that? Maybe automatically closing the parentheses? Maybe again, this could be checked for and give an easier error message? And it would be great to add this to the documentation (if I did not just overlook it).

## Reproducible code

```r
library(quanteda)
# Original regex: (die|der|den)( unteren)? 99( |-)?(%|prozent)
# Translated: the (lower )?99( |-)(%|percent)
# Easier: the (lower )?99 percent

dict <- dictionary(list(regexdict = c("the (lower )?99 percent")))
text <- c("This is the first document. It includes a the 99 percent.",
          "This is the second document. It includes the lower 99 percent.",
          "This is the third document. It only includes the upper 1 percent.")
text <- corpus(text)
tokens <- tokens(text)

tokens <- tokens_compound(x = tokens,
                          pattern = dict,
                          valuetype = "regex")
```

## Expected behavior

I thought that such a pattern would match both instances in the the example (which it would through normal regular expression search. But I just figured out that my approach could not really have worked as intended anyways. I think I cannot include optional words in a multi-word compound because with tokens_compound each single-word token has to be matched by the according part of the pattern. I will probably manually generate the compounds by replacing spaces with _ before tokenizing.

**Comment by koheiw** [comment]
Regex or glob patterns works only within token types, so `(lower)?` does not work to match sequences of tokens that do not have "lower". **quaneda** does not support variable length patterns, but we can always pass multiple patterns.  

``` r
library(quanteda)
#> Package version: 1.4.1
#> Parallel computing: 2 of 8 threads used.
#> See https://quanteda.io for tutorials and examples.
#> 
#> Attaching package: 'quanteda'
#> The following object is masked from 'package:utils':
#> 
#>     View
dict <- dictionary(list(regexdict = c("the lower 99 percent", "the 99 percent")))
text <- c("This is the first document. It includes a the 99 percent.",
          "This is the second document. It includes the lower 99 percent.",
          "This is the third document. It only includes the upper 1 percent.")
corp <- corpus(text)
toks <- tokens(corp)

tokens_compound(x = toks, pattern = dict)
#> tokens from 3 documents.
#> text1 :
#>  [1] "This"           "is"             "the"            "first"         
#>  [5] "document"       "."              "It"             "includes"      
#>  [9] "a"              "the_99_percent" "."             
#> 
#> text2 :
#>  [1] "This"                 "is"                   "the"                 
#>  [4] "second"               "document"             "."                   
#>  [7] "It"                   "includes"             "the_lower_99_percent"
#> [10] "."                   
#> 
#> text3 :
#>  [1] "This"     "is"       "the"      "third"    "document" "."       
#>  [7] "It"       "only"     "includes" "the"      "upper"    "1"       
#> [13] "percent"  "."
```
You can also do this to remove of "lower". 
```r
tokens_replace(toks, phrase("lower 99 percent"), phrase("99 percent"))
```

**Comment by jogrue** [comment]
Thank you. Sorry, I only figured that out while writing the bug report. I still think that the current behavior could at least return a better warning/message. Right now it leads people to look for non-matching parentheses in their patterns. At least, it took me quite some time to figure it out.

And for anyone interested, I started to write a few functions that replace phrases of varying length with tokens (phrases separated by "_"). I did not really test or use this, yet, and I have no idea how the code's performance will be: [https://gist.github.com/jogrue/6e91394c59afdd4a59cd333be9baf0ae](https://gist.github.com/jogrue/6e91394c59afdd4a59cd333be9baf0ae)
